### ISSUE TITLE
Type inference fails for nested async closure: "closure type is always closure"

### ISSUE THREAD

**Issue by mergeconflict** [rust-lang/rust-analyzer#21539]
Async closure inside sync closure fails type inference with error `closure type is always closure`

## Reproduction

```rust
fn bug() { || { async || {} }; }
```

## Notes

- `|| async {}` does NOT trigger the bug
- `async || {}` DOES trigger the bug
- The outer `||` is required

## Related

- [#15724](https://github.com/rust-lang/rust-analyzer/issues/15724)

**Comment by mergeconflict** [comment]
Standalone test harness:

```rs
// main.rs
use ra_ap_hir::Crate;
use ra_ap_hir_ty::attach_db;
use ra_ap_ide_db::RootDatabase;
use ra_ap_test_fixture::WithFixture;

fn main() {
    tracing_subscriber::fmt()
        .with_env_filter("warn,ra_ap_hir_ty=error")
        .with_writer(std::io::stderr)
        .init();

    let db = RootDatabase::with_files(
        r#"
//- /lib.rs crate:repro
fn bug() { || { async || {} }; }
"#,
    );

    attach_db(&db, || {
        for krate in Crate::all(&db) {
            for decl in krate.root_module(&db).declarations(&db) {
                // ModuleDef::diagnostics triggers body inference for functions
                let _ = decl.diagnostics(&db, false);
            }
        }
    });
}

```

**Comment by ChayimFriedman2** [comment]
I'm currently rewriting closure analysis so assigning this to myself.

**Comment by ChayimFriedman2** [comment]
Fixed now, after my rewrite.
